### PR TITLE
periodically remove expired entries from the cache

### PR DESCRIPTION
**Pull request by neelance**

### CODE REVIEW

**Comment by emidoots** [comment]
Nice, def. a step in the right direction. I think this should work great. Thanks for looking into this @neelance!

LGTM

File: grpccache.go [hunk 1]
```diff
@@ -164,6 +169,31 @@ func (c *Cache) Store(ctx context.Context, method string, arg proto.Message, res
 	return nil
 }
 
+// cleanOldEntries removes expired entries from the cache. It does so in a
+// separate goroutine and only once every 10 seconds. The method assumes that
+// the cache mutex is locked.
+func (c *Cache) cleanOldEntries() {
+	now := time.Now()
+	if now.After(c.lastClean.Add(10 * time.Second)) {
+		c.lastClean = now
```
Comment (dmitshur) [inline grpccache.go:178]: `lastClean` isn't protected by the mutex, but it looks like it should be. Is this okay?

Comment (emidoots) [inline grpccache.go:178]: @shurcooL the method is documented as `The method assumes that the cache mutex is locked.` and also is only called under `c.mu.Lock()` in both cases above. I think it's okay.

File: grpccache.go [hunk 2]
```diff
@@ -164,6 +169,31 @@ func (c *Cache) Store(ctx context.Context, method string, arg proto.Message, res
...
+		// clear old entries asynchronously
+		go func() {
+			c.mu.Lock()
```
Comment (beyang) [inline grpccache.go:181]: This holds a lock while it iterates through all cache entries, which would block every single gRPC request, right? This seems like a cheap operation, but should we upper bound the time somehow?

Also, cache exhaustion is still a possibility here. We should implement an actual eviction policy in the future. @neelance can you file an issue to track on the Trello? You can put it under tech debt and we'll make sure it's prioritized.

File: grpccache.go (hunk 1 above)
Comment (dmitshur) [inline grpccache.go:178]: I missed that. It looks good.

> The method assumes that the cache mutex is locked.

I've seen this better phrased as "The caller _must be_ holding the Cache.mu lock."

**Comment by beyang** [comment]
Merging and filing the follow up here: https://trello.com/c/XoEJMOZH/1126-add-cache-eviction-policy-to-grpc-cache

File: grpccache.go (hunk 1 above)
Comment (neelance) [inline grpccache.go:178]: @shurcooL I like that. Changed: https://github.com/sourcegraph/grpccache/commit/18bf8cab6fc00c3b9bbdcf7422eb6e85877c7d0c

**Comment by neelance** [comment]
@beyang Thanks for creating the Trello card, would have done it myself today. Cache eviction is a delicate thing, since as you mentioned you don't want to create CPU usage peaks or long-term locks. That's why I have decided to use a very simple approach right now so it does not introduce any complicated surprises. We may need to refine it in the future. But if we get to the situation that we need to add more logic, then I'd vote for using some existing implementation instead of trying to get this right on our own.
It is also not straightforward to say what should happen if the cache runs full. If it happens suddenly, then there is probably some strange request pattern (e.g. many requests to some endpoint) and letting that pattern purge everything else from the cache might not improve anything. Cache monitoring and warnings might be useful.

**Pull request merged**